File: frontend/types/tag.ts
export interface Tag {
  id: number;
  name: string;
  slug: string;
  description?: string;
  parent_id?: number;
  color?: string;
  created_at: string;
  updated_at?: string;
}

export interface TagTree extends Tag {
  children: TagTree[];
}

export interface TagCreate {
  name: string;
  slug: string;
  description?: string;
  parent_id?: number;
  color?: string;
}

export interface TagUpdate {
  name?: string;
  slug?: string;
  description?: string;
  parent_id?: number;
  color?: string;
}

File: backend/app/schemas/tag.py
from pydantic import BaseModel, Field
from typing import Optional, List
from datetime import datetime

class TagBase(BaseModel):
    name: str = Field(..., min_length=1, max_length=100)
    slug: str = Field(..., min_length=1, max_length=100)
    description: Optional[str] = Field(None, max_length=500)
    parent_id: Optional[int] = None
    color: Optional[str] = Field(None, pattern=r'^#[0-9A-Fa-f]{6}$')

class TagCreate(TagBase):
    pass

class TagUpdate(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=100)
    slug: Optional[str] = Field(None, min_length=1, max_length=100)
    description: Optional[str] = Field(None, max_length=500)
    parent_id: Optional[int] = None
    color: Optional[str] = Field(None, pattern=r'^#[0-9A-Fa-f]{6}$')

class TagResponse(TagBase):
    id: int
    created_at: datetime
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True

class TagTree(TagResponse):
    children: List['TagTree'] = []

    class Config:
        from_attributes = True

# Enable forward references
TagTree.model_rebuild()
